File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/DeletionUpdate.cpp
// FILE: DeletionUpdate.cpp /////////////////////////////////////////////////////////////////////////
// Author: Graham Smallwood, August 2002
// Desc:   Update that will count down a lifetime and destroy object when it reaches zero
///////////////////////////////////////////////////////////////////////////////////////////////////

// INCLUDES ///////////////////////////////////////////////////////////////////////////////////////
#include "PreRTS.h"
#include "Common/RandomValue.h"
#include "Common/Xfer.h"
#include "GameLogic/GameLogic.h"
#include "GameLogic/Object.h"
#include "GameLogic/Module/DeletionUpdate.h"

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
DeletionUpdate::DeletionUpdate( Thing *thing, const ModuleData* moduleData ) : UpdateModule( thing, moduleData )
{
	m_dieFrame = 0;
	const DeletionUpdateModuleData* d = getDeletionUpdateModuleData();
	UnsignedInt delay = calcSleepDelay(d->m_minFrames, d->m_maxFrames);
	setWakeFrame(getObject(), UPDATE_SLEEP(delay));
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
DeletionUpdate::~DeletionUpdate( void )
{
}


//#define CRISS_CROSS_GEOMETRY

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
void DeletionUpdate::setLifetimeRange( UnsignedInt minFrames, UnsignedInt maxFrames )
{
	
#if defined _DEBUG && defined CRISS_CROSS_GEOMETRY
	setWakeFrame(getObject(), UPDATE_SLEEP(2));
#else
	UnsignedInt delay = calcSleepDelay(minFrames, maxFrames);
	setWakeFrame(getObject(), UPDATE_SLEEP(delay));
#endif
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
UnsignedInt DeletionUpdate::calcSleepDelay(UnsignedInt minFrames, UnsignedInt maxFrames)
{
	UnsignedInt delay = GameLogicRandomValue( minFrames, maxFrames );
	if (delay < 1) delay = 1;
	m_dieFrame = TheGameLogic->getFrame() + delay;
	return delay;
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
UpdateSleepTime DeletionUpdate::update( void )
{
	// Destroy (NOT kill) if time is up
#if defined _DEBUG  && defined CRISS_CROSS_GEOMETRY
	Object *obj = getObject();
	if (obj)
	{
		GeometryInfo geom =	geom=obj->getGeometryInfo();
		geom.setMajorRadius(obj->getGeometryInfo().getMinorRadius());// CRIS
		geom.setMinorRadius(obj->getGeometryInfo().getMajorRadius());// CROSS
		obj->setGeometryInfo(geom);
	}

	if (TheGameLogic->getFrame() > m_dieFrame)
	{
		TheGameLogic->destroyObject( getObject() );
		return UPDATE_SLEEP_FOREVER;
	}


	return UPDATE_SLEEP(2);
#else
	TheGameLogic->destroyObject( getObject() );
	return UPDATE_SLEEP_FOREVER;
#endif
}

// ------------------------------------------------------------------------------------------------
/** CRC */
// ------------------------------------------------------------------------------------------------
void DeletionUpdate::crc( Xfer *xfer )
{

	// extend base class
	UpdateModule::crc( xfer );

}  // end crc

// ------------------------------------------------------------------------------------------------
/** Xfer method
	* Version Info:
	* 1: Initial version */
// ------------------------------------------------------------------------------------------------
void DeletionUpdate::xfer( Xfer *xfer )
{

	// version
	XferVersion currentVersion = 1;
	XferVersion version = currentVersion;
	xfer->xferVersion( &version, currentVersion );

	// extend base class
	UpdateModule::xfer( xfer );

	// die frame
	xfer->xferUnsignedInt( &m_dieFrame );

}  // end xfer

// ------------------------------------------------------------------------------------------------
/** Load post process */
// ------------------------------------------------------------------------------------------------
void DeletionUpdate::loadPostProcess( void )
{

	// extend base class
	UpdateModule::loadPostProcess();

}  // end loadPostProcess

File: GeneralsRust/Code/GameEngine/GameLogic/src/object/update/deletion_update.rs
// DeletionUpdate - Counts down a lifetime and destroys object when it reaches zero
// Author: Graham Smallwood, August 2002
// Ported to Rust

use crate::prelude::*;

#[derive(Debug, Clone)]
pub struct DeletionUpdateModuleData {
    pub min_frames: u32,
    pub max_frames: u32,
}

impl Default for DeletionUpdateModuleData {
    fn default() -> Self {
        Self {
            min_frames: 0,
            max_frames: 0,
        }
    }
}

#[derive(Debug, Clone)]
pub struct DeletionUpdate {
    thing: ThingId,
    #[allow(dead_code)]
    module_data: DeletionUpdateModuleData,
    die_frame: u32,
}

impl DeletionUpdate {
    pub fn new(thing: ThingId, module_data: DeletionUpdateModuleData) -> Self {
        let current_frame = crate::helpers::TheGameLogic::get_frame();
        let delay = Self::calc_sleep_delay(&module_data, current_frame);
        Self {
            thing,
            module_data,
            die_frame: delay,
        }
    }

    fn calc_sleep_delay(data: &DeletionUpdateModuleData, current_frame: u32) -> u32 {
        let mut delay = game_logic_random_value(data.min_frames, data.max_frames);
        if delay < 1 {
            delay = 1;
        }
        current_frame + delay
    }

    pub fn set_lifetime_range(&mut self, min_frames: u32, max_frames: u32, current_frame: u32) {
        self.die_frame = Self::calc_sleep_delay(
            &DeletionUpdateModuleData {
                min_frames,
                max_frames,
            },
            current_frame,
        );
    }

    pub fn update(&mut self, ctx: &mut UpdateContext<'_>) -> UpdateSleepTime {
        // C++ destroys whenever the scheduled update is invoked; timing is owned by the scheduler.
        if let Some(object) = ctx.game_logic.find_object(self.thing) {
            ctx.game_logic.destroy_object(object.id());
        }
        UpdateSleepTime::Forever
    }

    pub fn save(&self, xfer: &mut dyn Xfer) {
        let xfer_io = |result: std::io::Result<()>, field: &str| {
            if let Err(err) = result {
                panic!("DeletionUpdate::save failed to xfer {field}: {err}");
            }
        };

        xfer.xfer_version_write(1);
        let mut die_frame = self.die_frame;
        xfer_io(xfer.xfer_u32(&mut die_frame), "die_frame");
    }

    pub fn load(&mut self, xfer: &mut dyn Xfer) {
        let xfer_io = |result: std::io::Result<()>, field: &str| {
            if let Err(err) = result {
                panic!("DeletionUpdate::load failed to xfer {field}: {err}");
            }
        };

        let version = xfer.xfer_version_read();
        if version >= 1 {
            xfer_io(xfer.xfer_u32(&mut self.die_frame), "die_frame");
        }
    }
}

fn game_logic_random_value(min: u32, max: u32) -> u32 {
    if min >= max {
        return min;
    }
    crate::helpers::game_logic_random_value(min, max)
}
